src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,733 @@
+            }
+            return true;
+        }
+        if (isa<AddrSpaceCastInst>(inst) || isa<BitCastInst>(inst)) {
+            push_inst(inst);
+            return true;
+        }
+        if (auto gep = dyn_cast<GetElementPtrInst>(inst)) {
+            APInt apoffset(sizeof(void*) * 8, cur.offset, true);
+            if (ignore_tag && (!gep->accumulateConstantOffset(*DL, apoffset) ||
+                               apoffset.isNegative()))
+                ignore_tag = false;
+            push_inst(inst);
+            cur.offset = apoffset.getLimitedValue();
+            return true;
+        }
+        return false;
+    };
+
+    while (true) {
+        assert(cur.user_it != cur.user_end);
+        auto user = *cur.user_it;
+        auto inst = dyn_cast<Instruction>(user);
+        ++cur.user_it;
+        if (!inst)
+            return false;
+        if (!check_inst(inst))
+            return false;
+        uses.insert(inst);
+        if (cur.user_it == cur.user_end) {
+            if (stack.empty())
+                return true;
+            cur = stack.back();
+            stack.pop_back();
+        }
+    }
+}
+
+// Both arguments should be pointer of the same type but possibly different address spaces
+// `orig_i` is always in addrspace 0.
+// This function needs to handle all cases `AllocOpt::checkInst` can handle.
+// This function should not erase any safepoint so that the lifetime marker can find and cache
+// all the original safepoints.
+void AllocOpt::replaceUsesWith(Instruction *orig_inst, Instruction *new_inst,
+                               ReplaceUsesStack &stack)
+{
+    auto simple_replace = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (orig_i->user_empty()) {
+            if (orig_i != orig_inst)
+                orig_i->eraseFromParent();
+            return true;
+        }
+        Type *orig_t = orig_i->getType();
+        Type *new_t = new_i->getType();
+        if (orig_t == new_t) {
+            orig_i->replaceAllUsesWith(new_i);
+            if (orig_i != orig_inst)
+                orig_i->eraseFromParent();
+            return true;
+        }
+        return false;
+    };
+    if (simple_replace(orig_inst, new_inst))
+        return;
+    stack.clear();
+    stack.emplace_back(orig_inst, new_inst);
+    ReplaceUsesFrame *cur = &stack[0];
+    auto finish_cur = [&] () {
+        assert(cur->orig_i->user_empty());
+        if (cur->orig_i != orig_inst) {
+            cur->orig_i->eraseFromParent();
+        }
+    };
+    auto push_frame = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (simple_replace(orig_i, new_i))
+            return;
+        stack.emplace_back(orig_i, new_i);
+        cur = &stack.back();
+    };
+    auto replace_inst = [&] (Instruction *user) {
+        Instruction *orig_i = cur->orig_i;
+        Instruction *new_i = cur->new_i;
+        if (isa<LoadInst>(user) || isa<StoreInst>(user)) {
+            user->replaceUsesOfWith(orig_i, new_i);
```

Comment:
The type of `orig_i` and `new_i` are always the same ignoring addrspace. If they are not, they only differ in addrspace.